tests/unit/test_symmetric_crypto.cpp

```diff
@@ -0,0 +1,312 @@
+
+using EncrytDecryptTestData = std::tuple<SymmetricCipherMode, std::string>;
+
+static const std::vector<SymmetricCipherMode> AllSupportedCIpherModesToTest{
+        SymmetricCipherMode::CBC,
+        SymmetricCipherMode::GCM,
+        SymmetricCipherMode::CTR
+};
+
+class SymmetricCipherBase
+{
+public:
+    void SetUp()  {
+        _secretKey = utility::randomBytes(256 / 8);
+
+        std::string tmp;
+        for (auto i = 0; i < 1024 + 17; i++) {
+            tmp += "a string which will be concatenated multiple times-";
+        }
+        _plaintext = {tmp.begin(), tmp.end()};
+    }
+
+    std::vector<uint8_t> _secretKey;
+    std::vector<uint8_t> _plaintext;
+};
+
+class SymmetricCipherTest : public SymmetricCipherBase, public testing::TestWithParam<EncrytDecryptTestData>
+{
+    void SetUp() override  {
+        SymmetricCipherBase::SetUp();
+    }
+};
+
+class SymmetricCipherAdvancedTest : public SymmetricCipherBase, public testing::TestWithParam<SymmetricCipherMode>
+{
+    void SetUp() override  {
+        SymmetricCipherBase::SetUp();
+    }
+};
+
+
+TEST_P(SymmetricCipherTest, EncryptDecryptDifferentPlaintextLength)
+{
+    auto operationMode = std::get<0>(GetParam());
+    auto plaintextString = std::get<1>(GetParam());
+    std::vector<uint8_t> plaintext{plaintextString.begin(), plaintextString.end()};
+
+    // When: encrypt and then decrypt the plaintext
+    auto encryptoBuilder = AESCipherBuilder{operationMode, _secretKey};
+    std::shared_ptr<SymmetricCipher> encryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        encryptor = encryptoBuilder.buildAuthenticatedEncryptor();
+    }
+    else {
+        encryptor = encryptoBuilder.buildEncryptor();
+    }
+
+    encryptor->update(plaintext);
+    auto ciphertext = encryptor->finish();
+    auto iv = encryptor->getIV();
+
+    std::vector<uint8_t> tag;
+    auto authenticatedEncryptor = dynamic_cast<AuthenticatedEncryptionI*>(encryptor.get());
+    if (authenticatedEncryptor) {
+        tag = authenticatedEncryptor->getAuthTag();
+    }
+
+    auto decryptorBuilder = AESCipherBuilder{operationMode, _secretKey}.setIV(iv);
+    std::shared_ptr<SymmetricCipher> decryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        decryptor = encryptoBuilder.buildAuthenticatedDecryptor();
+    }
+    else {
+        decryptor = encryptoBuilder.buildDecryptor();
+    }
+
+    auto authenticatedDecryptor = dynamic_cast<AuthenticatedEncryptionI*>(decryptor.get());
+    if (authenticatedDecryptor) {
+        authenticatedDecryptor->setAuthTag(tag);
+    }
+
+    decryptor->update(ciphertext);
+    auto decryptedText = decryptor->finish();
+
+    // Then: Decrypted text should match the plaintext
+    ASSERT_THAT(decryptedText, ::testing::ElementsAreArray(plaintext));
+}
+
+
+static std::vector<EncrytDecryptTestData> prepareTestDataForMode(SymmetricCipherMode mode)
+{
+    static const std::vector<std::string> PlaintextStringsOfDifferentLength
```

Comment:
style: Usually constants are all caps with underscore in this lib.

Reply:
done